File: contracts-v1/helpers/merkle-tree.ts
import { StandardMerkleTree } from '@openzeppelin/merkle-tree';
import { BigNumber } from 'ethers';

export default interface Leaf {
  address: string;
  amount: BigNumber;
}

export function buildMerkleTree(leaves: Leaf[]): StandardMerkleTree<string[]> {
  const leavesArray = leaves.map(leaf => [leaf.address, leaf.amount.toString()]);
  return StandardMerkleTree.of(leavesArray, ['address', 'uint256']);
}

export function getProof(tree: StandardMerkleTree<string[]>, address: string): string[] {
  for (const [i, v] of tree.entries()) {
    if (v[0] === address) {
      return tree.getProof(i);
    }
  }
  throw Error('Address not found');
}

File: backend/app/modules/common/merkle_tree.py
from typing import List

from multiproof import StandardMerkleTree

from app.exceptions import MissingAddress
from app.infrastructure import database
from app.modules.dto import AccountFundsDTO

LEAF_ENCODING: List[str] = ["address", "uint256"]


def get_rewards_merkle_tree_for_epoch(epoch: int) -> StandardMerkleTree:
    leaves = [
        AccountFundsDTO(r.address, int(r.amount))
        for r in database.rewards.get_by_epoch(epoch)
    ]
    return build_merkle_tree(leaves)


def build_merkle_tree(
    leaves: List[AccountFundsDTO],
) -> StandardMerkleTree:
    return StandardMerkleTree.of(
        [[leaf.address, leaf.amount] for leaf in leaves], LEAF_ENCODING
    )


def get_proof(mt: StandardMerkleTree, address: str) -> List[str]:
    for i, leaf in enumerate(mt.values):
        if leaf.value[0] == address:
            return mt.get_proof(i)
    raise MissingAddress(address)
